File: sdk-js/src/admin-data-labels.ts
import type { RequestOptions } from "./index.js";

export type DataLabelSensitivity = "public" | "internal" | "personal" | "restricted";
export type DataLabelPropagationPolicy = "none" | "event_type_to_event" | "pack_selected";

export type DataLabelValue = {
  uuid: string;
  value: string;
  displayName: string;
  description: string;
  enabled: boolean;
  createdAt: string;
  updatedAt: string;
};
export type DataLabelDefinition = {
  uuid: string;
  authority: string;
  key: string;
  displayName: string;
  description: string;
  allowedScopes: string[];
  sensitivity: DataLabelSensitivity;
  intendedUse: string;
  synonyms: string[];
  propagationPolicy: DataLabelPropagationPolicy;
  enabled: boolean;
  createdAt: string;
  updatedAt: string;
  values: DataLabelValue[];
};
export type DataLabelDefinitionCreateRequest = Omit<
  DataLabelDefinition,
  "uuid" | "authority" | "enabled" | "createdAt" | "updatedAt" | "values"
>;
export type DataLabelDefinitionUpdateRequest = Omit<DataLabelDefinitionCreateRequest, "key">;
export type DataLabelValueCreateRequest = Pick<DataLabelValue, "value" | "displayName" | "description">;
export type DataLabelValueUpdateRequest = Pick<DataLabelValue, "displayName" | "description">;
export type DataLabelDefinitionListResponse = { definitions: DataLabelDefinition[] };
export type DataLabelUsage = {
  definitionUuid: string;
  valueUuid: string;
  eventTypeDefaults: number;
  schemaFieldAssignments: number;
  total: number;
};
export type DataLabelUsageListResponse = { usage: DataLabelUsage[] };
export type EventTypeDataLabelDefault = { eventTypeSlug: string; definitionUuid: string; valueUuid: string };
export type EventTypeDataLabelDefaultRequest = { valueUuid: string };
export type SchemaFieldDataLabelAssignment = {
  assignmentUuid: string;
  schemaUuid: string;
  fieldPath: string;
  definitionUuid: string;
  valueUuid: string;
};
export type SchemaFieldDataLabelAssignmentRequest = Omit<
  SchemaFieldDataLabelAssignment,
  "assignmentUuid" | "schemaUuid"
>;
export type DataLabelAssignmentListResponse = {
  eventTypeDefaults: EventTypeDataLabelDefault[];
  schemaFieldAssignments: SchemaFieldDataLabelAssignment[];
};
export type DescriptiveDataLabel = {
  definitionUuid: string;
  key: string;
  displayName: string;
  description?: string;
  sensitivity?: DataLabelSensitivity;
  intendedUse?: string;
  synonyms?: string[];
  enabled: boolean;
};
export type DataLabelCatalogue = { labels: DescriptiveDataLabel[] };
export type DataLabelCatalogueAssignment = {
  scope: string;
  eventTypeSlug?: string;
  schemaUuid?: string;
  fieldPath?: string;
  definitionUuid: string;
  valueUuid: string;
};
export type DataLabelCatalogueDataset = {
  displayName: string;
  unit?: string;
  definitionUuid?: string;
  valueUuids?: string[];
};
export type DataLabelCataloguePack = {
  displayName: string;
  enabled: boolean;
  datasets: DataLabelCatalogueDataset[];
};
export type DataLabelCatalogueResponse = {
  catalogue: DataLabelCatalogue;
  assignments: DataLabelCatalogueAssignment[];
  reportingPacks: DataLabelCataloguePack[];
  fingerprint: string;
};

type AdminRequester = <T>(method: string, path: string, body?: unknown, options?: RequestOptions) => Promise<T>;
const includeDisabledQuery = (include: boolean): string => (include ? "?includeDisabled=true" : "");

export class DataLabelAdminClient {
  constructor(private readonly request: AdminRequester) {}
  list(includeDisabled = false): Promise<DataLabelDefinitionListResponse> {
    return this.request("GET", `/data-labels${includeDisabledQuery(includeDisabled)}`);
  }
  catalogue(includeDisabled = false): Promise<DataLabelCatalogueResponse> {
    return this.request("GET", `/data-labels/catalogue${includeDisabledQuery(includeDisabled)}`);
  }
  get(uuid: string, includeDisabled = false): Promise<DataLabelDefinition> {
    return this.request("GET", `/data-labels/${encodeURIComponent(uuid)}${includeDisabledQuery(includeDisabled)}`);
  }
  create(body: DataLabelDefinitionCreateRequest): Promise<DataLabelDefinition> {
    return this.request("POST", "/data-labels", body);
  }
  update(uuid: string, body: DataLabelDefinitionUpdateRequest): Promise<DataLabelDefinition> {
    return this.request("PATCH", `/data-labels/${encodeURIComponent(uuid)}`, body);
  }
  disable(uuid: string): Promise<void> {
    return this.request("POST", `/data-labels/${encodeURIComponent(uuid)}/disable`);
  }
  createValue(uuid: string, body: DataLabelValueCreateRequest): Promise<DataLabelValue> {
    return this.request("POST", `/data-labels/${encodeURIComponent(uuid)}/values`, body);
  }
  updateValue(uuid: string, body: DataLabelValueUpdateRequest): Promise<DataLabelValue> {
    return this.request("PATCH", `/data-label-values/${encodeURIComponent(uuid)}`, body);
  }
  disableValue(uuid: string): Promise<void> {
    return this.request("POST", `/data-label-values/${encodeURIComponent(uuid)}/disable`);
  }
  listUsage(): Promise<DataLabelUsageListResponse> {
    return this.request("GET", "/data-labels/usage");
  }
  listAssignments(): Promise<DataLabelAssignmentListResponse> {
    return this.request("GET", "/data-label-assignments");
  }
  setEventTypeDefault(slug: string, definitionUuid: string, body: EventTypeDataLabelDefaultRequest): Promise<void> {
    return this.request(
      "PUT",
      `/event-types/${encodeURIComponent(slug)}/data-label-defaults/${encodeURIComponent(definitionUuid)}`,
      body,
    );
  }
  removeEventTypeDefault(slug: string, definitionUuid: string): Promise<void> {
    return this.request(
      "DELETE",
      `/event-types/${encodeURIComponent(slug)}/data-label-defaults/${encodeURIComponent(definitionUuid)}`,
    );
  }
  setSchemaFieldAssignment(schemaUuid: string, body: SchemaFieldDataLabelAssignmentRequest): Promise<void> {
    return this.request("PUT", `/event-schemas/${encodeURIComponent(schemaUuid)}/field-data-labels`, body);
  }
  removeSchemaFieldAssignment(assignmentUuid: string): Promise<void> {
    return this.request("DELETE", `/data-label-assignments/schema-fields/${encodeURIComponent(assignmentUuid)}`);
  }
}

File: sdk-python/src/custd/admin_data_labels.py
from __future__ import annotations

from typing import NotRequired, TypedDict, cast

from .client import AdminClient, quote_path


class DataLabelValue(TypedDict):
    uuid: str
    value: str
    displayName: str
    description: str
    enabled: bool
    createdAt: str
    updatedAt: str


class DataLabelDefinition(TypedDict):
    uuid: str
    authority: str
    key: str
    displayName: str
    description: str
    allowedScopes: list[str]
    sensitivity: str
    intendedUse: str
    synonyms: list[str]
    propagationPolicy: str
    enabled: bool
    createdAt: str
    updatedAt: str
    values: list[DataLabelValue]


class DataLabelDefinitionCreateRequest(TypedDict):
    key: str
    displayName: str
    description: str
    allowedScopes: list[str]
    sensitivity: str
    intendedUse: str
    synonyms: list[str]
    propagationPolicy: str


class DataLabelDefinitionUpdateRequest(TypedDict):
    displayName: str
    description: str
    allowedScopes: list[str]
    sensitivity: str
    intendedUse: str
    synonyms: list[str]
    propagationPolicy: str


class DataLabelValueCreateRequest(TypedDict):
    value: str
    displayName: str
    description: str


class DataLabelValueUpdateRequest(TypedDict):
    displayName: str
    description: str


class EventTypeDataLabelDefaultRequest(TypedDict):
    valueUuid: str


class SchemaFieldDataLabelAssignmentRequest(TypedDict):
    fieldPath: str
    definitionUuid: str
    valueUuid: str


class DataLabelDefinitionListResponse(TypedDict):
    definitions: list[DataLabelDefinition]


class DataLabelUsage(TypedDict):
    definitionUuid: str
    valueUuid: str
    eventTypeDefaults: int
    schemaFieldAssignments: int
    total: int


class DataLabelUsageListResponse(TypedDict):
    usage: list[DataLabelUsage]


class EventTypeDataLabelDefault(TypedDict):
    eventTypeSlug: str
    definitionUuid: str
    valueUuid: str


class SchemaFieldDataLabelAssignment(TypedDict):
    assignmentUuid: str
    schemaUuid: str
    fieldPath: str
    definitionUuid: str
    valueUuid: str


class DataLabelAssignmentListResponse(TypedDict):
    eventTypeDefaults: list[EventTypeDataLabelDefault]
    schemaFieldAssignments: list[SchemaFieldDataLabelAssignment]


class DescriptiveDataLabel(TypedDict):
    definitionUuid: str
    key: str
    displayName: str
    description: NotRequired[str]
    sensitivity: NotRequired[str]
    intendedUse: NotRequired[str]
    synonyms: NotRequired[list[str]]
    enabled: bool


class DataLabelCatalogue(TypedDict):
    labels: list[DescriptiveDataLabel]


class DataLabelCatalogueAssignment(TypedDict):
    scope: str
    eventTypeSlug: NotRequired[str]
    schemaUuid: NotRequired[str]
    fieldPath: NotRequired[str]
    definitionUuid: str
    valueUuid: str


class DataLabelCatalogueDataset(TypedDict):
    displayName: str
    unit: NotRequired[str]
    definitionUuid: NotRequired[str]
    valueUuids: NotRequired[list[str]]


class DataLabelCataloguePack(TypedDict):
    displayName: str
    enabled: bool
    datasets: list[DataLabelCatalogueDataset]


class DataLabelCatalogueResponse(TypedDict):
    catalogue: DataLabelCatalogue
    assignments: list[DataLabelCatalogueAssignment]
    reportingPacks: list[DataLabelCataloguePack]
    fingerprint: str


class DataLabelAdminClient:
    def __init__(self, admin: AdminClient) -> None:
        self._admin = admin

    @staticmethod
    def _include_disabled(include: bool) -> str:
        return "?includeDisabled=true" if include else ""

    def list(self, include_disabled: bool = False) -> DataLabelDefinitionListResponse:
        return cast(
            DataLabelDefinitionListResponse,
            self._admin.request("GET", "/data-labels" + self._include_disabled(include_disabled)),
        )

    def catalogue(self, include_disabled: bool = False) -> DataLabelCatalogueResponse:
        return cast(
            DataLabelCatalogueResponse,
            self._admin.request("GET", "/data-labels/catalogue" + self._include_disabled(include_disabled)),
        )

    def get(self, uuid: str, include_disabled: bool = False) -> DataLabelDefinition:
        return cast(
            DataLabelDefinition,
            self._admin.request("GET", f"/data-labels/{quote_path(uuid)}{self._include_disabled(include_disabled)}"),
        )

    def create(self, body: DataLabelDefinitionCreateRequest) -> DataLabelDefinition:
        return cast(DataLabelDefinition, self._admin.request("POST", "/data-labels", dict(body)))

    def update(self, uuid: str, body: DataLabelDefinitionUpdateRequest) -> DataLabelDefinition:
        return cast(DataLabelDefinition, self._admin.request("PATCH", f"/data-labels/{quote_path(uuid)}", dict(body)))

    def disable(self, uuid: str) -> None:
        self._admin.request("POST", f"/data-labels/{quote_path(uuid)}/disable")

    def create_value(self, definition_uuid: str, body: DataLabelValueCreateRequest) -> DataLabelValue:
        return cast(
            DataLabelValue,
            self._admin.request("POST", f"/data-labels/{quote_path(definition_uuid)}/values", dict(body)),
        )

    def update_value(self, value_uuid: str, body: DataLabelValueUpdateRequest) -> DataLabelValue:
        return cast(
            DataLabelValue, self._admin.request("PATCH", f"/data-label-values/{quote_path(value_uuid)}", dict(body))
        )

    def disable_value(self, value_uuid: str) -> None:
        self._admin.request("POST", f"/data-label-values/{quote_path(value_uuid)}/disable")

    def list_usage(self) -> DataLabelUsageListResponse:
        return cast(DataLabelUsageListResponse, self._admin.request("GET", "/data-labels/usage"))

    def list_assignments(self) -> DataLabelAssignmentListResponse:
        return cast(DataLabelAssignmentListResponse, self._admin.request("GET", "/data-label-assignments"))

    def set_event_type_default(self, slug: str, definition_uuid: str, body: EventTypeDataLabelDefaultRequest) -> None:
        self._admin.request(
            "PUT", f"/event-types/{quote_path(slug)}/data-label-defaults/{quote_path(definition_uuid)}", dict(body)
        )

    def remove_event_type_default(self, slug: str, definition_uuid: str) -> None:
        self._admin.request(
            "DELETE", f"/event-types/{quote_path(slug)}/data-label-defaults/{quote_path(definition_uuid)}"
        )

    def set_schema_field_assignment(self, schema_uuid: str, body: SchemaFieldDataLabelAssignmentRequest) -> None:
        self._admin.request("PUT", f"/event-schemas/{quote_path(schema_uuid)}/field-data-labels", dict(body))

    def remove_schema_field_assignment(self, assignment_uuid: str) -> None:
        self._admin.request("DELETE", f"/data-label-assignments/schema-fields/{quote_path(assignment_uuid)}")
